File: blast/blast-stress-solver/src/rapier/splitMigrator.ts
import type { SplitChild } from "./types";

export type ExistingBodyState = {
  handle: number;
  nodeIndices: ReadonlySet<number>;
  isFixed: boolean;
};

export type PlannerChild = SplitChild & { index: number };

export type SplitMigrationPlan = {
  reuse: Array<{ childIndex: number; bodyHandle: number }>;
  create: Array<{ childIndex: number }>;
};

export function planSplitMigration(
  bodies: ExistingBodyState[],
  children: PlannerChild[],
  timings?: { onDuration?: (ms: number) => void },
): SplitMigrationPlan {
  const start = typeof performance !== "undefined" ? performance.now() : Date.now();

  if (bodies.length === 0 || children.length === 0) {
    const plan = {
      reuse: [],
      create: children.map((child) => ({ childIndex: child.index })),
    };
    timings?.onDuration?.(
      (typeof performance !== "undefined" ? performance.now() : Date.now()) - start,
    );
    return plan;
  }

  const bodyHashes = buildBodyHashIndex(bodies);
  const reuse: Array<{ childIndex: number; bodyHandle: number }> = [];
  const reusedChildren = new Set<number>();
  const assignedBodies = new Set<number>();
  const unmatchedChildren: PlannerChild[] = [];

  for (const child of children) {
    const hash = hashNodes(child.nodes);
    const candidates = bodyHashes.get(hash);
    if (!candidates) {
      unmatchedChildren.push(child);
      continue;
    }
    const body = candidates.find(
      (candidate) =>
        !assignedBodies.has(candidate.handle) &&
        (candidate.isFixed ? child.isSupport : true),
    );
    if (!body) {
      unmatchedChildren.push(child);
      continue;
    }
    reuse.push({ childIndex: child.index, bodyHandle: body.handle });
    reusedChildren.add(child.index);
    assignedBodies.add(body.handle);
  }

  const unmatchedBodies = bodies.filter((body) => !assignedBodies.has(body.handle));

  let remainingPlan: SplitMigrationPlan = { reuse: [], create: [] };
  if (unmatchedBodies.length > 0 && unmatchedChildren.length > 0) {
    remainingPlan = planViaHungarian(unmatchedBodies, unmatchedChildren);
  } else {
    remainingPlan.create = unmatchedChildren.map((child) => ({
      childIndex: child.index,
    }));
  }

  const plan = {
    reuse: [...reuse, ...remainingPlan.reuse],
    create: remainingPlan.create,
  };
  timings?.onDuration?.(
    (typeof performance !== "undefined" ? performance.now() : Date.now()) - start,
  );
  return plan;
}

function planViaHungarian(
  bodies: ExistingBodyState[],
  children: PlannerChild[],
): SplitMigrationPlan {
  const overlapMatrix = buildOverlapMatrix(bodies, children);
  const assignments = hungarianMax(overlapMatrix);
  const reuse: Array<{ childIndex: number; bodyHandle: number }> = [];
  const reusedChildren = new Set<number>();

  bodies.forEach((body, rowIdx) => {
    const assignedCol = assignments[rowIdx];
    if (
      assignedCol == null ||
      assignedCol < 0 ||
      assignedCol >= children.length
    ) {
      return;
    }
    const overlap = overlapMatrix[rowIdx][assignedCol] ?? 0;
    if (overlap > 0) {
      reuse.push({
        childIndex: children[assignedCol].index,
        bodyHandle: body.handle,
      });
      reusedChildren.add(children[assignedCol].index);
    }
  });

  const create = children
    .filter((child) => !reusedChildren.has(child.index))
    .map((child) => ({ childIndex: child.index }));
  return { reuse, create };
}

function buildBodyHashIndex(
  bodies: ExistingBodyState[],
): Map<string, ExistingBodyState[]> {
  const map = new Map<string, ExistingBodyState[]>();
  for (const body of bodies) {
    const hash = hashNodes(body.nodeIndices);
    const bucket = map.get(hash);
    if (bucket) {
      bucket.push(body);
    } else {
      map.set(hash, [body]);
    }
  }
  return map;
}

function hashNodes(nodes: Iterable<number>): string {
  const list = Array.isArray(nodes)
    ? nodes.slice()
    : Array.from(nodes);
  list.sort((a, b) => a - b);
  return list.join(",");
}

function buildOverlapMatrix(
  bodies: ExistingBodyState[],
  children: PlannerChild[],
): number[][] {
  return bodies.map((body) => {
    return children.map((child) => {
      if (body.isFixed && !child.isSupport) {
        return 0;
      }
      let overlap = 0;
      for (const nodeIndex of child.nodes) {
        if (body.nodeIndices.has(nodeIndex)) overlap += 1;
      }
      return overlap;
    });
  });
}

function hungarianMax(matrix: number[][]): number[] {
  const rows = matrix.length;
  const cols = matrix[0]?.length ?? 0;
  if (!rows || !cols) return new Array(rows).fill(-1);
  const size = Math.max(rows, cols);
  let maxVal = 0;
  for (const row of matrix) {
    for (const value of row) maxVal = Math.max(maxVal, value);
  }
  const cost = Array.from({ length: size }, (_, i) =>
    Array.from({ length: size }, (_, j) => {
      const value = matrix[i]?.[j] ?? 0;
      return maxVal - value;
    }),
  );
  const assignment = hungarian(cost);
  return assignment.slice(0, rows);
}

function hungarian(cost: number[][]): number[] {
  const n = cost.length;
  const m = cost[0]?.length ?? 0;
  const size = Math.max(n, m);
  const paddedCost = Array.from({ length: size }, (_, i) =>
    Array.from({ length: size }, (_, j) => cost[i]?.[j] ?? 0),
  );
  const u = new Array(size + 1).fill(0);
  const v = new Array(size + 1).fill(0);
  const p = new Array(size + 1).fill(0);
  const way = new Array(size + 1).fill(0);

  for (let i = 1; i <= size; i++) {
    p[0] = i;
    const minv = new Array(size + 1).fill(Infinity);
    const used = new Array(size + 1).fill(false);
    let j0 = 0;
    do {
      used[j0] = true;
      const i0 = p[j0];
      let delta = Infinity;
      let j1 = 0;
      for (let j = 1; j <= size; j++) {
        if (used[j]) continue;
        const cur = paddedCost[i0 - 1][j - 1] - u[i0] - v[j];
        if (cur < minv[j]) {
          minv[j] = cur;
          way[j] = j0;
        }
        if (minv[j] < delta) {
          delta = minv[j];
          j1 = j;
        }
      }
      for (let j = 0; j <= size; j++) {
        if (used[j]) {
          u[p[j]] += delta;
          v[j] -= delta;
        } else {
          minv[j] -= delta;
        }
      }
      j0 = j1;
    } while (p[j0] !== 0);
    do {
      const j1 = way[j0];
      p[j0] = p[j1];
      j0 = j1;
    } while (j0);
  }

  const result = new Array(size).fill(-1);
  for (let j = 1; j <= size; j++) {
    if (p[j] > 0) {
      result[p[j] - 1] = j - 1;
    }
  }
  return result;
}

File: blast/blast-stress-solver-rs/src/rapier/split_migrator.rs
use std::collections::{HashMap, HashSet};

use rapier3d::prelude::RigidBodyHandle;

use crate::types::SplitChild;

/// Body state before a split.
pub struct ExistingBodyState {
    pub handle: RigidBodyHandle,
    pub node_indices: HashSet<u32>,
    pub is_fixed: bool,
}

/// Plan for how to handle split children: reuse existing bodies or create new ones.
pub struct SplitMigrationPlan {
    /// Children that can reuse an existing body (same node set).
    pub reuse: Vec<ReuseEntry>,
    /// Children that need a new body created.
    pub create: Vec<CreateEntry>,
}

pub struct ReuseEntry {
    pub child_index: usize,
    pub body_handle: RigidBodyHandle,
}

pub struct CreateEntry {
    pub child_index: usize,
}

#[derive(Clone, Copy, Debug, Default)]
pub struct PlannerChildSupport {
    pub is_support: bool,
}

/// Plan body reuse vs. creation for split children.
///
/// Uses hash-based matching: if a child's node set exactly matches an existing body,
/// reuse that body. Otherwise, create a new one. Largest children are prioritized.
pub fn plan_split_migration(
    bodies: &[ExistingBodyState],
    children: &[SplitChild],
) -> SplitMigrationPlan {
    plan_split_migration_with_support(
        bodies,
        children,
        &vec![PlannerChildSupport::default(); children.len()],
    )
}

pub fn plan_split_migration_with_support(
    bodies: &[ExistingBodyState],
    children: &[SplitChild],
    child_support: &[PlannerChildSupport],
) -> SplitMigrationPlan {
    if bodies.is_empty() || children.is_empty() {
        return SplitMigrationPlan {
            reuse: Vec::new(),
            create: (0..children.len())
                .map(|i| CreateEntry { child_index: i })
                .collect(),
        };
    }

    // Build hash index of existing bodies
    let mut body_hashes: HashMap<u64, Vec<usize>> = HashMap::new();
    for (i, body) in bodies.iter().enumerate() {
        let hash = hash_node_set(&body.node_indices);
        body_hashes.entry(hash).or_default().push(i);
    }

    let mut reuse = Vec::new();
    let mut assigned_bodies = HashSet::new();
    let mut unmatched_children = Vec::new();

    for (ci, child) in children.iter().enumerate() {
        let child_set: HashSet<u32> = child.nodes.iter().copied().collect();
        let hash = hash_node_set(&child_set);

        let mut matched = false;
        if let Some(candidates) = body_hashes.get(&hash) {
            for &bi in candidates {
                if assigned_bodies.contains(&bi) {
                    continue;
                }
                if bodies[bi].is_fixed
                    && !child_support
                        .get(ci)
                        .copied()
                        .unwrap_or_default()
                        .is_support
                {
                    continue;
                }
                if bodies[bi].node_indices == child_set {
                    reuse.push(ReuseEntry {
                        child_index: ci,
                        body_handle: bodies[bi].handle,
                    });
                    assigned_bodies.insert(bi);
                    matched = true;
                    break;
                }
            }
        }
        if !matched {
            unmatched_children.push(ci);
        }
    }

    let unmatched_bodies: Vec<usize> = bodies
        .iter()
        .enumerate()
        .filter_map(|(idx, _)| (!assigned_bodies.contains(&idx)).then_some(idx))
        .collect();

    if !unmatched_bodies.is_empty() && !unmatched_children.is_empty() {
        let overlap = build_overlap_matrix(
            bodies,
            &unmatched_bodies,
            children,
            &unmatched_children,
            child_support,
        );
        let assignments = hungarian_max(&overlap);
        let mut reused_children = HashSet::new();

        for (row_idx, assignment) in assignments.into_iter().enumerate() {
            let Some(col_idx) = assignment else {
                continue;
            };
            let overlap_score = overlap
                .get(row_idx)
                .and_then(|row| row.get(col_idx))
                .copied()
                .unwrap_or(0);
            if overlap_score == 0 {
                continue;
            }

            let body_idx = unmatched_bodies[row_idx];
            let child_index = unmatched_children[col_idx];
            reuse.push(ReuseEntry {
                child_index,
                body_handle: bodies[body_idx].handle,
            });
            reused_children.insert(child_index);
        }

        let create = unmatched_children
            .into_iter()
            .filter(|child_index| !reused_children.contains(child_index))
            .map(|child_index| CreateEntry { child_index })
            .collect();
        return SplitMigrationPlan { reuse, create };
    }

    let create = unmatched_children
        .into_iter()
        .map(|child_index| CreateEntry { child_index })
        .collect();
    SplitMigrationPlan { reuse, create }
}

fn hash_node_set(nodes: &HashSet<u32>) -> u64 {
    let mut sorted: Vec<u32> = nodes.iter().copied().collect();
    sorted.sort_unstable();
    let mut hash = 0u64;
    for n in sorted {
        // Simple FNV-1a-like hash
        hash ^= n as u64;
        hash = hash.wrapping_mul(0x100000001b3);
    }
    hash
}

fn build_overlap_matrix(
    bodies: &[ExistingBodyState],
    unmatched_bodies: &[usize],
    children: &[SplitChild],
    unmatched_children: &[usize],
    child_support: &[PlannerChildSupport],
) -> Vec<Vec<usize>> {
    unmatched_bodies
        .iter()
        .map(|&body_idx| {
            let body = &bodies[body_idx];
            unmatched_children
                .iter()
                .map(|&child_idx| {
                    if body.is_fixed
                        && !child_support
                            .get(child_idx)
                            .copied()
                            .unwrap_or_default()
                            .is_support
                    {
                        return 0;
                    }

                    children[child_idx]
                        .nodes
                        .iter()
                        .filter(|node| body.node_indices.contains(node))
                        .count()
                })
                .collect()
        })
        .collect()
}

fn hungarian_max(matrix: &[Vec<usize>]) -> Vec<Option<usize>> {
    let rows = matrix.len();
    let cols = matrix.first().map(Vec::len).unwrap_or(0);
    if rows == 0 || cols == 0 {
        return vec![None; rows];
    }

    let size = rows.max(cols);
    let max_val = matrix
        .iter()
        .flat_map(|row| row.iter())
        .copied()
        .max()
        .unwrap_or(0) as i64;
    let mut cost = vec![vec![max_val; size]; size];
    for (i, row) in matrix.iter().enumerate() {
        for (j, value) in row.iter().enumerate() {
            cost[i][j] = max_val - (*value as i64);
        }
    }

    let assignments = hungarian(&cost);
    assignments
        .into_iter()
        .take(rows)
        .map(|col| (col < cols).then_some(col))
        .collect()
}

fn hungarian(cost: &[Vec<i64>]) -> Vec<usize> {
    let size = cost.len();
    let mut u = vec![0i64; size + 1];
    let mut v = vec![0i64; size + 1];
    let mut p = vec![0usize; size + 1];
    let mut way = vec![0usize; size + 1];

    for i in 1..=size {
        p[0] = i;
        let mut minv = vec![i64::MAX; size + 1];
        let mut used = vec![false; size + 1];
        let mut j0 = 0usize;
        loop {
            used[j0] = true;
            let i0 = p[j0];
            let mut delta = i64::MAX;
            let mut j1 = 0usize;
            for j in 1..=size {
                if used[j] {
                    continue;
                }
                let cur = cost[i0 - 1][j - 1] - u[i0] - v[j];
                if cur < minv[j] {
                    minv[j] = cur;
                    way[j] = j0;
                }
                if minv[j] < delta {
                    delta = minv[j];
                    j1 = j;
                }
            }
            for j in 0..=size {
                if used[j] {
                    u[p[j]] += delta;
                    v[j] -= delta;
                } else if minv[j] != i64::MAX {
                    minv[j] -= delta;
                }
            }
            j0 = j1;
            if p[j0] == 0 {
                break;
            }
        }
        loop {
            let j1 = way[j0];
            p[j0] = p[j1];
            j0 = j1;
            if j0 == 0 {
                break;
            }
        }
    }

    let mut result = vec![usize::MAX; size];
    for j in 1..=size {
        if p[j] > 0 {
            result[p[j] - 1] = j - 1;
        }
    }
    result
}
